### ISSUE TITLE
Unicode characters are not preserved.

### ISSUE THREAD

**Issue by ccrusius** [lspitzner/brittany#127]
`brittany` discards Unicode symbols that GHC supports. For example, it will format something like

```haskell
f :: A → B → C
```

as

```haskell
f :: A -> B -> C
```

**Comment by lspitzner** [comment]
I don't use that extension, but it would certainly be nice to have it supported in brittany.

Related additional features might also include some flags/commands to refactor/convert between unicode and non-unicode versions. (Current behaviour is like `const False`; you ask for `id` and we might want to allow the user to choose between those two and `const True` applied to "unicodeness".) I am not sure how much unicode syntax there is or how consistently its users apply it.

**Comment by ccrusius** [comment]
I agree - it would be nice if the "Unicoding" could be automatically done by `brittany`.

As far to "how much," `brittany` already keeps Unicode characters for a bunch of things. My guess would be that it is currently screwing up the ones in the [UnicodeSyntax extension](https://www.schoolofhaskell.com/school/to-infinity-and-beyond/pick-of-the-week/guide-to-ghc-extensions/basic-syntax-extensions#unicodesyntax) list, probably because those are the only ones hardcoded in GHC itself (as opposed to being provided by other packages).

**Comment by ccrusius** [comment]
So I just wrote a patch that makes `brittany` automatically transform `UnicodeSyntax` tokens into their Unicode equivalent whenever `-XUnicodeSyntax` is passed as a GHC option, which I think is reasonable. I'll think about a pull request - it requires some paperwork at work.

**Comment by ccrusius** [comment]
Now for a little bit more detail, in case those who are more familiar with the code have better ideas: the problem is that `brittany`, when building function signatures, expressions, and such, writes down hardcoded strings for the following tokens: `::`, `->`, `forall`, `<-`, and `=>`. It does that in various places in the following files:

* `src/Language/Haskell/Brittany/Internal/Layouters/Decl.hs`
* `src/Language/Haskell/Brittany/Internal/Layouters/Expr.hs`
* `src/Language/Haskell/Brittany/Internal/Layouters/Pattern.hs`
* `src/Language/Haskell/Brittany/Internal/Layouters/Stmt.hs`
* `src/Language/Haskell/Brittany/Internal/Layouters/Type.hs`

Changing the hardcoded strings to parameterized ones is not a big deal, it took me about an hour, and that was the first time I saw the code.

But a better idea would be to keep the proper tokens. I think the AST has the information there (for example, `AnnDcolonU` for Unicode, versus `AnnDcolon` for ASCII). That requires diving into the code to an extent I'll probably won't find the time to do. Hopefully it is not too hard for those who know the code inside-out!

**Comment by lspitzner** [comment]
> But a better idea would be to keep the proper tokens. I think the AST has the information there (for example, AnnDcolonU for Unicode, versus AnnDcolon for ASCII).

Right, and brittany already has the utils to query for the presence of those annotations for relevant nodes (via `hasAnnKeyword`). For example, see [This test for the `AnnSimpleQuote` annotation thingy](https://github.com/lspitzner/brittany/blob/1330aeb6b4d3a3138bca89e1f3ee966677ee93db/src/Language/Haskell/Brittany/Internal/LayouterBasics.hs#L234).

You need to be sure to run this test on the correct node, but by using `--dump-ast-full` this should not be too hard to get right.

**Comment by mb720** [comment]
If someone would like to contribute a patch to make Brittany keep the Unicode syntax (`→, ∷, ←, etc.`), where in the code would that someone best start?
